validator/src/journal/candidate_block.rs

```diff
@@ -431,10 +428,17 @@ impl CandidateBlock {
                     committed_txn_cache.add_batch(&batch.clone());
                 }
             } else {
+                self.invalid_batches.push(batch.clone());
```

Comment:
Isn't the contents of this just the `bad_batches` vector?  If so, it seems like you are missing the above case when dependencies have failed.

Reply:
Additionally, you could just store the batch header signature, and not the whole batch (which could be sizeable).

Reply:
I was hesitant to use the `bad_batches` vector _because_ of the dependencies. I could imagine a case where you'd want to retry the batches with unmet dependencies but not those that are invalid for any other reason.

Just using headers is definitely more efficient. Good catch.